### ISSUE TITLE
Undertrained yolo v8 model for object detcetion

### ISSUE THREAD

**Issue by mlruns** [ultralytics/ultralytics#4622]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I trained a YOLOv8 medium model to detect bolts. My dataset has 294 images. I used a 90-10 train-val split. When I made a prediction on an test image by drawing a circle using paint on same image,  the model predicted it as a bolt with 90% accuracy. I believe this is because the model is underfit.

To eloborate:
I took an image from my dataset and draw a circle similar as bolt with paint brush and then made a prediction on modified image.

I would like to understand how to read these graphs and what conclusion to make from it. 

My results looks like this.

![image](https://github.com/ultralytics/ultralytics/assets/131228873/cb7a8136-44fe-4d4d-bc52-40c82400e6b1)

### Additional

...................

**Comment by ExtReMLapin** [comment]
You want to dead the recall, precision curves.

Recall = out of 100 sheeps, how much sheep do I detect ? 1.0 = 100 sheeps, 0.9 = 90 sheeps

Precision = out of the 100 detected sheeps, how many are really sheeps ?

**Comment by mlruns** [comment]
> You want to dead the recall, precision curves.
> 
> Recall = out of 100 sheeps, how much sheep do I detect ? 1.0 = 100 sheeps, 0.9 = 90 sheeps
> 
> Precision = out of the 100 detected sheeps, how many are really sheeps ?

what i understood by dead reacll and precsision is getting R/P is geeting close to constant as number of epochs increasing. 

Something like this:

![image](https://github.com/ultralytics/ultralytics/assets/131228873/805762d5-954f-449e-b38c-ce06728f7ebd)

**Comment by ExtReMLapin** [comment]
Right now you have very good metrics, but map50-95 is still increasing so it could be worth the wait to let it train a little more till it overfits.

When it overfits, the curve will go down, as seen on this picture

![image](https://github.com/ultralytics/ultralytics/assets/3909752/6988c2c5-7423-4d2c-b8c5-939c5fa0739b)

**Comment by mlruns** [comment]
Thank you for explaining the graph and giving me clear picture.

I trained the model for 500 epochs, but early stopping stopped it after 200 epochs. I am training model with deafult hyperparameters. Do you think changing the early stopping parameter value will help the map50-95 curve ? 

current early stopping value = 50.

**Comment by ExtReMLapin** [comment]
Early stopping ? You mean patience ? Personally I always disabled it.

**Comment by ExtReMLapin** [comment]
Oopsie woopsie, i actually didn't read your very first message entirely.

So the issue it , when you present something to your model that it has never seen before (a circle drawn with a tool), it think it's a bolt.

You could need more "background" images, in other words, images that do not contain any of the objects you try to detect.

You need to have a dataset that is "representative" of what you will have in reality/at runtime. Will you ever have a circle visible on the pictures ? if yes, add it to your dataset using background images (image with the circle but not annotated, to make understand the model, there is nothing to detect here), if no, you can ignore the false positive

**Comment by glenn-jocher** [comment]
@mlruns yes, you are on the right track. 

When a model encounters something that it has not been trained on, as you correctly identified with the circle drawn with a paint tool, it may misclassify it because it tries to fit the new information within its learned parameters. 

The variety and diversity of your dataset play a crucial role in your model's performance. If the circle is something that will be present in real-time applications and should not be identified as a bolt, including such scenarios in your training data can be beneficial. 

Include more background images - images that do not contain any bolts but have similar features to a bolt, like the circle. This can help the model learn the difference between a bolt and similar-looking objects in the background. If the circle will not appear in the real-time application, you can generally ignore such false positives.

Remember to balance your dataset with both target objects and background images to avoid the model becoming biased. Create a dataset representative of your real-world application to ensure the model learns to deal with similar scenarios effectively.

**Comment by mlruns** [comment]
@ExtReMLapin  @glenn-jocher Thankyou for great explaniantion!

There will not be any images like this in real time application. I guess I can ignore thses false positive for now.

on a related note, is it a underfitted model if it is not able to diffrentiate between bolt or circle? I got this comment on my model performace that this is underfitted model as it not able to learn bolt features correctly.

**Comment by glenn-jocher** [comment]
@ExtReMLapin it's excellent to hear that you found our explanations helpful, and that there wouldn't be images similar to this in real-time applications. Yes, in that case, you can ignore these false positives for now.

On your query regarding whether the model is underfitted if it's not able to differentiate between a bolt and a circle, while it might point to certain limitations in the model, it might not necessarily reflect underfitting per se.

Underfitting refers to a scenario where a machine learning model is not able to capture the underlying patterns of the data due to its oversimplified nature. This often results in poor performance on both the training and validation sets. A traditional symptom of underfitting is that both the training loss and validation loss are high or the training accuracy is low.

Inability to distinguish a bolt from a circle might, for instance, indicate that the model hasn't been trained on diverse representations of a bolt (i.e, seen enough examples of 'bolts' vs 'not bolts'). 

The important point to remember is that underfitting is generally a characteristic of the model not learning enough from the training data, and not necessarily about making mistakes on specific categories or types of objects.

I hope this clarifies things for you. Should you have more queries or require further clarification, feel free to ask.
